templates/extensions/the-graph/packages/subgraph/scripts/abi_copy.ts

```diff
@@ -0,0 +1,112 @@
+import * as fs from "fs";
+import chalk from "chalk";
+const parseAndCorrectJSON = (input: string): any => {
+  // Add double quotes around keys
+  let correctedJSON = input.replace(/(\w+)(?=\s*:)/g, '"$1"');
+
+  // Remove trailing commas
+  correctedJSON = correctedJSON.replace(/,(?=\s*[}\]])/g, "");
+
+  try {
+    return JSON.parse(correctedJSON);
+  } catch (error) {
+    console.error("Failed to parse JSON", error);
+    throw new Error("Failed to parse JSON");
+  }
+};
+
+type Contract = {
+  address: string;
+  abi: any[];
+};
+
+const GRAPH_DIR = "./";
+
+function publishContract(
+  contractName: string,
+  contractObject: Contract,
+  networkName: string
+) {
+  try {
+    const graphConfigPath = `${GRAPH_DIR}/networks.json`;
+    let graphConfig = "{}";
+    try {
+      if (fs.existsSync(graphConfigPath)) {
+        graphConfig = fs.readFileSync(graphConfigPath).toString();
+      }
+    } catch (e) {
+      console.log(e);
+    }
+
+    let graphConfigObject = JSON.parse(graphConfig);
+    if (!(networkName in graphConfigObject)) {
+      graphConfigObject[networkName] = {};
+    }
+    if (!(contractName in graphConfigObject[networkName])) {
+      graphConfigObject[networkName][contractName] = {};
+    }
+    graphConfigObject[networkName][contractName].address =
+      contractObject.address;
+
+    fs.writeFileSync(
+      graphConfigPath,
+      JSON.stringify(graphConfigObject, null, 2)
+    );
+    if (!fs.existsSync(`${GRAPH_DIR}/abis`)) fs.mkdirSync(`${GRAPH_DIR}/abis`);
+    fs.writeFileSync(
+      `${GRAPH_DIR}/abis/${networkName}_${contractName}.json`,
+      JSON.stringify(contractObject.abi, null, 2)
+    );
+
+    return true;
+  } catch (e) {
+    console.log(
+      "Failed to publish " + chalk.red(contractName) + " to the subgraph."
+    );
+    console.log(e);
+    return false;
+  }
+}
+
+const DEPLOYED_CONTRACTS_FILE = "../nextjs/contracts/deployedContracts.ts";
+async function main() {
+  const fileContent = fs.readFileSync(DEPLOYED_CONTRACTS_FILE, "utf8");
+
+  const pattern = /const deployedContracts = ({[^;]+}) as const;/s;
+  const match = fileContent.match(pattern);
+
+  if (!match || !match[1]) {
+    throw new Error(
+      `Failed to find deployedContracts in the ${DEPLOYED_CONTRACTS_FILE}`
+    );
+  }
+  const jsonString = match[1];
+
+  // Parse the JSON string
+  const deployedContracts = parseAndCorrectJSON(jsonString);
+  const localContracts = deployedContracts[31337];
```

Comment:
I have updated this file to work with both foundry and hardhat. 

Previously we were getting contract name, address, abi from hardhat deployment and hardhat path was hardcoded [checkout this](https://github.com/scaffold-eth/scaffold-eth-2/blob/041f25c9cc3d41b920fa75ff4ed90e9fc9d53ce3/packages/subgraph/scripts/abi_copy.ts#L5) (which won't work with foundry) . 

So now, instead of relying on hardhat deployment or foundry deployments (kind of hard to grab / track deployments in foundry) we now rely on frontend [`deployedContracts.ts`](https://github.com/scaffold-eth/scaffold-eth-2/blob/subgraph-package/packages/nextjs/contracts/deployedContracts.ts) file. 

The frontend `deployedContracts.ts` file will always be in sync with _local_ deployments. (Whenever people run `yarn deploy` both in hardhat, foundry this fill be updated with contract name, address, abi) 

For more details about data structure of file [checkout this](https://docs.scaffoldeth.io/hooks/)

### The Big assumption here is:

 This script is only used while shipping to local subgraph (which in turn means working with local network contracts 31337 chains) 
 
Should `yarn abi-copy` take in consideration for other chains too ? 

^ Currently it only grabs deployment for `31337` (hardhat / anvil) chain 

Please let me know if we should update it take in to consideration all chains deployments, just kept hardcoded `31337` for simplicity cc @kmjones1979

Reply:
Created #28, so this won't be a issue anymore